pkg/storaged/lvol-tabs.jsx

```diff
@@ -49,57 +48,145 @@ function lvol_rename(lvol) {
                         Value: lvol.Size,
+                        Min: lvol.Size,
                         Max: (pool ?
                               pool.Size * 3 :
                               lvol.Size + vgroup.FreeSize),
                         AllowInfinite: !!pool,
                         Round: vgroup.ExtentSize
-                      },
-                      { CheckBox: "fsys",
-                        Title: _("Resize Filesystem"),
-                        Value: block && block.IdUsage == "filesystem",
-                        visible: function () {
-                            return lvol.Type == "block";
-                        }
                       }
                   ],
                   Action: {
-                      Title: _("Resize"),
+                      Title: _("Grow"),
                       action: function (vals) {
+                          return utils.teardown_active_usage(client, usage).
+                                       then(function () {
+                                           return lvol_and_fsys_resize(client, lvol, vals.size,
+                                                                       info.grow_needs_unmount);
+                                       });
+                      }
+                  }
+    });
+}
 
-                          function error(msg) {
-                              return $.Deferred().reject({ message: msg }).promise();
-                          }
+function lvol_shrink(client, lvol, info) {
+    var block = client.lvols_block[lvol.path];
+    var vgroup = client.vgroups[lvol.VolumeGroup];
 
-                          var fsys = (block && block.IdUsage == "filesystem");
-                          if (!fsys && vals.size < lvol.Size)
-                              return error(_("This logical volume cannot be made smaller."));
+    var usage = utils.get_active_usage(client, block && info.shrink_needs_unmount? block.path : null);
 
-                          var options = { };
-                          if (fsys)
-                              options.resize_fsys = { t: 'b', v: fsys };
+    if (usage.Blocking) {
+        dialog.open({ Title: cockpit.format(_("$0 is in active use"), lvol.Name),
+                      Blocking: usage.Blocking,
+                      Fields: [ ]
+        });
+        return;
+    }
 
-                          return lvol.Resize(vals.size, options);
+    dialog.open({ Title: _("Shrink Logical Volume"),
+                  Teardown: usage.Teardown,
+                  Fields: [
+                      { SizeSlider: "size",
+                        Title: _("Size"),
+                        Value: lvol.Size,
+                        Max: lvol.Size,
+                        Round: vgroup.ExtentSize
+                      }
+                  ],
+                  Action: {
+                      Title: _("Shrink"),
+                      Danger: info.shrink_danger? _("Shrinking this volume might destroy data.") : null,
```

Comment:
IMHO this is too unspecific. I assume this means data *on this LV*, but the tests below point out a case where this is shown if the LV has no file system - i. e. there is no data to be destroyed on this LV. Also, the "might" is unclear - what does it depend on? Hopefully something more specific than "sheer luck"?

In which situations would a user realistically see this? Can we detect "safe" ones (like an LV without a file system) and not show it then, and disable shrinking in the unsafe cases?

Reply:
The "might" refers to the fact that Cockpit doesn't know whether there is data in the logical volume that the user cares about, or not.  There is no way to label a block device as empty, I think, and an empty block device is thus unfortunately the same as one with unrecognized content, as far as Cockpit is concerned. (One can use a empty partition table, I guess but they don't work well on device-mapper devices.)

A typical situation is a newly created logical volume that hasn't yet been formatted with a filesystem.

But this is a huge wart.  I'll try to learn more about this, maybe there is something we can do.

Reply:
The usual way is to ask blkid, i. e. let udev do the file system (or partition table) detection and export it to udisks. If the admin went through the trouble of calling wipefs on a previously used file system to make it "undetectable", then I think it's safe to assume that she doesn't care about that data any more. No other tool issues warnings like that, and IMHO it would just scare users who deal with block devices without file systems.

Reply:
Are you concerned about using the block device for raw data that gets exported to VMs, and the like, where blkid doesn't recognize anything? If the block device is opened for writing by some process already, it would be great if trying to resize it fails due to "busy", but that might not always be the case. But in that case the string could be less daunting. I have a hard time to come up with something which isn't utterly long and technical, as that appears to be such a small corner case..

Reply:
One short-lived example was VDO: for some time blkid didn't recognize the VDO signatures.

As discussed elsewhere, let's go back to the previous position: Unrecognized data can not be shrunk with Cockpit, and the user needs to delete the volume and recreate it with a smaller size explicitly.

Reply:
AFAICS `.danger` now is not set anywhere, so this is now obsolete.

Reply:
Right, good catch.  Fixed.